File: mobile/android/air/SubModules/NativeEnclave/src/main/java/org/mytonwallet/app_air/native_enclave/auth/PasscodeAuth.java
package org.mytonwallet.app_air.native_enclave.auth;

import androidx.fragment.app.FragmentActivity;

import org.mytonwallet.app_air.native_enclave.crypto.AesGcm;
import org.mytonwallet.app_air.native_enclave.crypto.HardwareKeyManager;
import org.mytonwallet.app_air.native_enclave.crypto.KeyDerivation;
import org.mytonwallet.app_air.native_enclave.storage.EnclaveStorage;

import javax.crypto.AEADBadTagException;
import javax.crypto.SecretKey;

public class PasscodeAuth implements EnclaveAuth {

    private final EnclaveStorage storage;

    public PasscodeAuth(EnclaveStorage storage) {
        this.storage = storage;
    }

    @Override
    public void setup(FragmentActivity activity, byte[] masterKey, String passcode, SetupCallback callback) {
        try {
            byte[] salt = KeyDerivation.generateSalt();
            SecretKey kek = KeyDerivation.deriveKeyFromPasscode(passcode, salt);
            String kekEncrypted = AesGcm.encrypt(masterKey, kek);

            storage.storePasscodeCredential(salt, kekEncrypted);
            callback.onSuccess();
        } catch (Exception e) {
            callback.onError("setupPasscode: " + e.getMessage());
        }
    }

    @Override
    public void authorize(FragmentActivity activity, String passcode, AuthorizeCallback callback) {
        try {
            byte[] salt = storage.loadSalt();
            if (salt == null) {
                throw new Exception("Passcode not configured");
            }
            String kekEncrypted = storage.loadMasterKey(AuthType.PASSCODE);
            if (kekEncrypted == null) {
                throw new Exception("Passcode not configured");
            }

            SecretKey kek = KeyDerivation.deriveKeyFromPasscode(passcode, salt);
            callback.onSuccess(AesGcm.decrypt(kekEncrypted, kek));
        } catch (AEADBadTagException e) {
            callback.onError(null);
        } catch (Exception e) {
            callback.onError("authorize: " + e.getMessage());
        }
    }

    @Override
    public void destroy() {
        HardwareKeyManager.deleteKey();
        storage.removeMasterKey(AuthType.PASSCODE);
        storage.removeSalt();
    }
}

File: mobile/ios/Air/SubModules/NativeEnclave/auth/PasscodeAuth.swift
import Foundation

final class PasscodeAuth: EnclaveAuth {
    private let storage: EnclaveStorage

    init(storage: EnclaveStorage) {
        self.storage = storage
    }

    func setup(masterKey: Data, passcode: String?) async throws {
        guard let passcode else {
            throw EnclaveError.passcodeRequired
        }

        let salt = try KeyDerivation.generateSalt()
        let kek = try KeyDerivation.derivePasscodeKey(passcode: passcode, salt: salt)
        let kekEncrypted = try AesGcm.encrypt(masterKey, keyData: kek)
        let hwEncrypted = try HardwareKeyManager.encrypt(kekEncrypted)

        try storage.storePasscodeCredential(
            PasscodeCredential(
                salt: salt,
                wrappedMasterKey: hwEncrypted
            )
        )
    }

    func authorize(passcode: String?) async throws -> Data {
        guard let passcode else {
            throw EnclaveError.passcodeRequired
        }
        guard let credential = try storage.loadPasscodeCredential() else {
            throw EnclaveError.passcodeNotConfigured
        }

        do {
            let kekEncrypted = try HardwareKeyManager.decrypt(credential.wrappedMasterKey)
            let kek = try KeyDerivation.derivePasscodeKey(
                passcode: passcode,
                salt: credential.salt
            )
            return try AesGcm.decrypt(kekEncrypted, keyData: kek)
        } catch {
            throw EnclaveError.invalidSessionToken
        }
    }

    func destroy() async {
        HardwareKeyManager.deleteDataKey()
        storage.removePasscodeCredential()
    }
}
